charms/worker/k8s/src/charm.py

```diff
@@ -372,41 +372,59 @@ def _get_valid_annotations(self) -> Optional[dict]:
 
-    def _configure_annotations(self, config: BootstrapConfig):
-        """Configure the annotations for the Canonical Kubernetes cluster.
+    def _get_cluster_config_from_charm_config(self) -> UserFacingClusterConfig:
+        """Retrieve the cluster config from charm configuration.
+
+        Returns:
+            UserFacingClusterConfig: The expected cluster configuration.
+        """
+        local_storage = LocalStorageConfig(
+            enabled=self.config.get("local_storage_enabled"),
+            local_path=self.config.get("local_storage_local_path"),
+            reclaim_policy=self.config.get("local_storage_reclaim_policy"),
+            set_default=self.config.get("local_storage_set_default"),
+        )
+
+        return UserFacingClusterConfig(
+            local_storage=local_storage,
+            annotations=self._get_valid_annotations(),
+        )
+
+    def _configure_cluster_config(self, config: BootstrapConfig):
+        """Configure the cluster config for the Canonical Kubernetes cluster.
 
         Args:
             config (BootstrapConfig): Configuration object to bootstrap the cluster.
         """
-        annotations = self._get_valid_annotations()
-        if annotations is None:
-            return
-
         if not config.cluster_config:
-            config.cluster_config = UserFacingClusterConfig(annotations=annotations)
+            config.cluster_config = self._get_cluster_config_from_charm_config()
         else:
-            config.cluster_config.annotations = annotations
+            charm_cluster_config = self._get_cluster_config_from_charm_config()
+            # Merge the charm and user cluster configurations
+            # The existing configuration takes precedence over the charm configuration as
+            # this information is retrieved elsewhere (e.g. cloud-provider).
+            merged_data = {**charm_cluster_config.dict(), **config.cluster_config.dict()}
+            config.cluster_config = UserFacingClusterConfig(**merged_data)
 
     @status.on_error(
-        ops.BlockedStatus("Invalid Annotations"),
+        ops.BlockedStatus("Invalid Cluster configuration"),
         AssertionError,
     )
-    def _update_annotations(self):
-        """Update the annotations for the Canonical Kubernetes cluster."""
-        annotations = self._get_valid_annotations()
-        if annotations is None:
-            return
+    def _update_cluster_config(self):
+        """Update the cluster configuration for the Canonical Kubernetes cluster."""
+        status.add(ops.MaintenanceStatus("Updating cluster configuration"))
+        log.info("Updating cluster configuration")
 
-        status.add(ops.MaintenanceStatus("Updating Annotations"))
-        log.info("Updating Annotations")
+        cluster_config = self._get_cluster_config_from_charm_config()
 
-        if self._stored.annotations == annotations:
+        if self._stored.cluster_config == cluster_config:
             return
```

Comment:
Yeah, this isn't very good.  I don't like that we're storing snap config in the charm.  It can get off sync. 

we should fetch the cluster config from the k8s snap, and compare it with the new charm config.  If they differ we need to update. 

Storing this info in the charm's stored state isn't the right path.

Reply:
updated